libvast/builtins/connectors/file.cpp

```diff
@@ -0,0 +1,203 @@
+        return caf::make_error(
+          ec::invalid_argument,
+          fmt::format("unexpected argument for 'file' connector: {}", arg));
+      }
+    }
+    if (path.empty()) {
+      return caf::make_error(ec::syntax_error,
+                             fmt::format("no file specified"));
+    }
+    auto fd = STDIN_FILENO;
+    if (is_socket) {
+      if (path == stdin_path) {
+        return caf::make_error(ec::filesystem_error, "cannot use STDIN as UNIX "
+                                                     "domain socket");
+      }
+      auto uds = detail::unix_domain_socket::connect(path);
+      if (!uds) {
+        return caf::make_error(
+          ec::filesystem_error,
+          fmt::format("Unable to connect to UNIX domain socket at {}:", path));
+      }
+      fd = uds.fd;
+    } else {
+      if (path != stdin_path) {
+        fd = ::open(path.c_str(), std::ios_base::binary | std::ios_base::in);
+        if (fd == -1) {
+          return caf::make_error(ec::filesystem_error,
+                                 fmt::format("open(2) for file {} failed {}:",
+                                             path, std::strerror(errno)));
+        }
+      }
+    }
+    return std::invoke(
+      [](auto timeout, auto fd, auto following) -> generator<chunk_ptr> {
+        auto in_buf = detail::fdinbuf(fd, max_chunk_size);
+        in_buf.read_timeout() = timeout;
+        auto current_data = std::vector<std::byte>{};
+        current_data.reserve(max_chunk_size);
+        auto eof_reached = false;
+        while (following or not eof_reached) {
+          auto current_char = in_buf.sbumpc();
+          if (current_char != detail::fdinbuf::traits_type::eof()) {
+            current_data.emplace_back(static_cast<std::byte>(current_char));
+          }
+          if (current_char == detail::fdinbuf::traits_type::eof()
+              or current_data.size() == max_chunk_size) {
+            eof_reached = (current_char == detail::fdinbuf::traits_type::eof()
+                           and not in_buf.timed_out());
+            if (eof_reached and current_data.empty() and not following) {
+              break;
+            }
+            auto chunk = chunk::make(std::exchange(current_data, {}));
+            co_yield std::move(chunk);
+            if (eof_reached and not following) {
+              break;
+            }
+            if (not eof_reached or following) {
+              current_data.reserve(max_chunk_size);
+            }
+          }
+        }
+        if (fd != STDIN_FILENO) {
+          ::close(fd);
+        }
+        co_return;
+      },
+      read_timeout, fd, following);
+  }
+
+  auto default_parser(std::span<std::string const> args) const
+    -> std::pair<std::string, std::vector<std::string>> override {
+    for (std::string_view arg : args) {
+      if (arg == "-" || arg == "stdin") {
+        break;
+      }
+      if (!arg.starts_with("-")) {
+        return {detail::file_path_to_parser(arg), {}};
+      }
```

Comment:
This does not work for `load stdin --timeout 1s foo.csv`, right?

Reply:
It shouldn't, but it does. I added a missing argument check in https://github.com/tenzir/vast/pull/3088/commits/26c17b32e7ba53a27f5027a6ab918f74b40d215c

Reply:
Sorry, I meant `from file --timeout 1s foo.csv`. This would try to infer the parser from `1s`, which leads to `[from_read.cpp:29] from/read expanded to 'load file --timeout 1s foo.csv | parse json '`. As a solution, you could refactor the argument parsing into a separate function and parse the arguments again here (and this confirms that our plugin API could really be improved).

Reply:
I added a simpler loop in https://github.com/tenzir/vast/pull/3088/commits/e0bd40fcdb71cacdd1cc8c363179674085fabe96 -  when the API is improved, this will be changed anyway.